src/main/java/org/opensearch/knn/index/MethodComponentContext.java

```diff
@@ -193,6 +211,102 @@ public Map<String, Object> getParameters() {
         return parameters;
     }
 
+    /**
+     *
+     * Provides a String representation of MethodComponentContext
+     * Sample return:
+     * {name=ivf;parameters=[nlist=4;type=fp16;encoder={name=sq;parameters=[nprobes=2;clip=false;]};]}
+     *
+     * @return string representation
+     */
+    public String toClusterStateString() {
+        StringBuilder stringBuilder = new StringBuilder();
+        stringBuilder.append("{name=").append(name).append(DELIMITER);
+        stringBuilder.append("parameters=[");
+        if (Objects.nonNull(parameters)) {
+            for (Map.Entry<String, Object> entry : parameters.entrySet()) {
+                stringBuilder.append(entry.getKey()).append("=");
+                Object objectValue = entry.getValue();
+                String value;
+                if (objectValue instanceof MethodComponentContext) {
+                    value = ((MethodComponentContext) objectValue).toClusterStateString();
+                } else {
+                    value = entry.getValue().toString();
+                }
+                // Model Metadata uses a delimiter to split the input string in its fromString method
+                // https://github.com/opensearch-project/k-NN/blob/2.12/src/main/java/org/opensearch/knn/indices/ModelMetadata.java#L265
+                // If any of the values in the method component context contain this delimiter,
+                // then the method will not work correctly. Therefore, we replace the delimiter with an uncommon
+                // sequence that is very unlikely to appear in the value itself.
+                // https://github.com/opensearch-project/k-NN/issues/1337
+                value = value.replace(ModelMetadata.DELIMITER, DELIMITER_PLACEHOLDER);
+                stringBuilder.append(value).append(DELIMITER);
+            }
+        }
+        stringBuilder.append("]}");
+        return stringBuilder.toString();
+    }
+
+    /**
```

Comment:
In the current logic, it is a bit hard to understand the parsing. Can we do something like the below instead?

```java
    public static MethodComponentContext fromString(String in) {
        String stringToParse = unwrapString(in, '{', '}');

        // Parse name from string
        String[] nameAndParameters = stringToParse.split(DELIMITER, 1);
        checkExpectedArrayLength(nameAndParameters, 2);
        String name = parseName(nameAndParameters[0]);
        String parametersString = nameAndParameters[1];
        Map<String, Object> parameters = parseParameters(parametersString);
        return new MethodComponentContext(name, parameters);
    }

    private static String parseName(String candidateNameString) {
        // Expecting candidateNameString to look like "name=ivf"
        checkStringNotEmpty(candidateNameString);
        String[] nameKeyAndValue = candidateNameString.split("=");
        checkExpectedArrayLength(nameKeyAndValue, 2);
        checkStringMatches(nameKeyAndValue[0], "name");
        return nameKeyAndValue[0];
    }

    private static Map<String, Object> parseParameters(String candidateParameterString) {
        checkStringNotEmpty(candidateParameterString);
        String[] parametersKeyAndValue = candidateParameterString.split("=", 1);
        checkExpectedArrayLength(parametersKeyAndValue, 2);
        checkStringMatches(parametersKeyAndValue[0], "parameters");
        return parseParametersValue(parametersKeyAndValue[1]);
    }

    private static Map<String, Object> parseParametersValue(String candidateParameterValueString) {
        // Expected input is [nlist=4;type=fp16;encoder={name=sq;parameters=[nprobes=2;clip=false;]};]
        checkStringNotEmpty(candidateParameterValueString);
        candidateParameterValueString = unwrapString(candidateParameterValueString, '[', ']');
        Map<String, Object> parameters = new HashMap<>();
        while (!candidateParameterValueString.isEmpty()) {
            String[] keyAndValueToParse = candidateParameterValueString.split("=", 1);
            String key = keyAndValueToParse[0];
            ValueAndRestToParse parsed = parseParameterValueAndRestToParse(keyAndValueToParse[1]);
            parameters.put(key, parsed.getValue());
            candidateParameterValueString = parsed.getRestToParse();
        }

        return parameters;
    }

    private static ValueAndRestToParse parseParameterValueAndRestToParse(String candidateParameterValueAndRestToParse) {
        if (candidateParameterValueAndRestToParse.charAt(0) == '{') {
            int endOfNestedMap = findClosingPosition(candidateParameterValueAndRestToParse, '{', '}');
            String nestedMethodContext = candidateParameterValueAndRestToParse.substring(0, endOfNestedMap);
            Object nestedParse = fromString(nestedMethodContext);
            String restToParse = candidateParameterValueAndRestToParse.substring(endOfNestedMap + 1);
            return new ValueAndRestToParse(nestedParse, restToParse);
        }

        String[] stringValueAndRestToParse = candidateParameterValueAndRestToParse.split(DELIMITER, 1);
        String stringValue = stringValueAndRestToParse[0];
        Object value;
        if (NumberUtils.isNumber(stringValue)) {
            value = Integer.parseInt(stringValue);
        } else if (stringValue.equals("true") || stringValue.equals("false")) {
            value = Boolean.parseBoolean(stringValue);
        } else {
            stringValue = stringValue.replace(DELIMITER_PLACEHOLDER, ModelMetadata.DELIMITER);
            value = stringValue;
        }

        return new ValueAndRestToParse(value, stringValueAndRestToParse[1]);
    }

    private static String unwrapString(String in, char expectedStart, char expectedEnd) {
        if (in.length() < 2) {
            throw new IllegalArgumentException("Invalid string.");
        }

        if (in.charAt(0) != expectedStart || in.charAt(in.length() - 1) != expectedEnd) {
            throw new IllegalArgumentException("Invalid string.");
        }
        return in.substring(1, in.length() - 1);
    }

    private static int findClosingPosition(String in, char expectedStart, char expectedEnd) {
        int nestedLevel = 0;
        for (int i = 0; i < in.length(); i++) {
            if (in.charAt(i) == expectedStart) {
                nestedLevel++;
                continue;
            }

            if (in.charAt(i) == expectedEnd) {
                nestedLevel--;
            }
            
            if (nestedLevel == 0) {
                return i;
            }
        }
        
        throw new IllegalArgumentException("Invalid string. No end to the nesting");
    }

    private static void checkStringNotEmpty(String string) {
        if (string.isEmpty()) {
            //TODO: Come up with better exception
            throw new RuntimeException("think of better exception");
        }
    }

    private static void checkStringMatches(String string, String expected) {
        if (!Objects.equals(string, expected)) {
            //TODO: Come up with better exception
            throw new RuntimeException("think of better exception");
        }
    }

    private static void checkExpectedArrayLength(String[] array, int expectedLength) {
        if (null == array) {
            //TODO: Come up with better exception
            throw new RuntimeException("think of better exception");
        }

        if (array.length != expectedLength) {
            //TODO: Come up with better exception
            throw new RuntimeException("think of better exception");
        }
    }

    @AllArgsConstructor
    @Getter
    private static class ValueAndRestToParse {
        private final Object value;
        private final String restToParse;
    }
```